# beliveaulab

Can we improve OligoMiner by implementing a method to interrogate tandem repeats in the genome? 

More than half of the human genome is comprised of repetitive sequences such as tandem repeats. These regions have been implicated in 3D genome organization and the maintenance of genome stability. Although these regions have established roles in the regulation of long noncoding genes, studying local repetitive elements has remained a technical challenge.Tandem repeats are not accessible with many conventional assays and these regions are not mappable which poses substantial roadblocks for visualization and mapping applications. FISH applications enables researchers to visualize the distribution of RNA and DNA in fixed samples. However more recent optimizations in synthetic biology has since provided researchers with the opportunity to construct FISH probes from oligonucleotides (oligos). Oligo-based FISH is a powerful single cell technique that allows researchers to control aspects of probe design such as binding energy and specificity. Here, we present TigeRFISH - a computational pipeline that allows for the design of optimal oligo FISH probes for tandem repeats genome wide. We implement existing computational tools including Tandem Repeat Finder coupled with OligoMiner to conduct k-mer based enrichment analysis to assess chromosome specificity followed by thermodynamic simulations of each probe. With TigeRFISHâ€™s capabilities to identify target repetitive candidates, we proceed with high-throughput microscopy at the single cell level to gain novel insight of the role these repetitive regions play in genome organization. Furthermore, the application of this technology as a part of the OligoMiner workflow may be used more broadly to facilitate the design of hybridization probes for a variety of applications genome-wide. 
